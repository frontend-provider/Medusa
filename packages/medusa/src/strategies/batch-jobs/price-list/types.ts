import { EntityManager } from "typeorm"
import { FileService } from "medusa-interfaces"

import {
  BatchJobService,
  PriceListService,
  ProductVariantService,
  RegionService,
} from "../../../services"
import { CsvSchema } from "../../../interfaces/csv-parser"
import { BatchJob } from "../../../models"

export type PriceListImportBatchJob = BatchJob & {
  context: PriceListImportJobContext
  result: Pick<BatchJob, "result"> & {
    operations: {
      [K in keyof typeof OperationType]: number
    }
  }
}

/**
 * DI props for the Product import strategy
 */
export type InjectedProps = {
  priceListService: PriceListService
  batchJobService: BatchJobService
  productVariantService: ProductVariantService
  regionService: RegionService
  fileService: typeof FileService
  manager: EntityManager
}

/**
 * Data shape returned by the CSVParser.
 */
export type TParsedPriceListImportRowData = Record<
  string,
  string | number | (string | number | object)[]
>

export type PriceListImportOperationPrice = {
  region_id?: string
  currency_code: string
  amount: number
}

export type PriceListImportOperation = {
  variant_id: string
  prices: PriceListImportOperationPrice[]
}

export type ParsedPriceListImportPrice =
  | {
      amount: number
      currency_code: string
    }
  | {
      amount: number
      region_name: string
    }

/**
 * CSV parser's row reducer result data shape.
 */
export type TBuiltPriceListImportLine = Record<string, any>

/**
 * Schema definition of for an import CSV file.
 */
export type PriceListImportCsvSchema = CsvSchema<
  TParsedPriceListImportRowData,
  TBuiltPriceListImportLine
>

/**
 * Import Batch job context column type.
 */
export type PriceListImportJobContext = {
  price_list_id: string
  fileKey: string
}

/**
 * Supported batch job import ops.
 */
export enum OperationType {
  PricesCreate = "PRICE_LIST_PRICE_CREATE",
}
