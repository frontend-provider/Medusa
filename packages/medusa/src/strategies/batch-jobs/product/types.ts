import { EntityManager } from "typeorm"
import { FileService } from "medusa-interfaces"

import {
  BatchJobService,
  ProductService,
  ProductVariantService,
  RegionService,
  SalesChannelService,
  ShippingProfileService,
} from "../../../services"
import { CsvSchema } from "../../../interfaces/csv-parser"
import { FlagRouter } from "../../../utils/flag-router"
import { BatchJob } from "../../../models"

export type ProductImportBatchJob = BatchJob & {
  result: Pick<BatchJob, "result"> & {
    operations: {
      [K in keyof typeof OperationType]: number
    }
  }
}

/**
 * DI props for the Product import strategy
 */
export type InjectedProps = {
  batchJobService: BatchJobService
  productService: ProductService
  productVariantService: ProductVariantService
  shippingProfileService: ShippingProfileService
  salesChannelService: SalesChannelService
  regionService: RegionService
  fileService: typeof FileService

  featureFlagRouter: FlagRouter
  manager: EntityManager
}

/**
 * Data shape returned by the CSVParser.
 */
export type TParsedProductImportRowData = Record<
  string,
  string | number | (string | number | object)[]
>

/**
 * CSV parser's row reducer result data shape.
 */
export type TBuiltProductImportLine = Record<string, any>

/**
 * Schema definition of for an import CSV file.
 */
export type ProductImportCsvSchema = CsvSchema<
  TParsedProductImportRowData,
  TBuiltProductImportLine
>

/**
 * Import Batch job context column type.
 */
export type ImportJobContext = {
  total: number
  fileKey: string
}

/**
 * Supported batch job import ops.
 */
export enum OperationType {
  ProductCreate = "PRODUCT_CREATE",
  ProductUpdate = "PRODUCT_UPDATE",
  VariantCreate = "VARIANT_CREATE",
  VariantUpdate = "VARIANT_UPDATE",
}
