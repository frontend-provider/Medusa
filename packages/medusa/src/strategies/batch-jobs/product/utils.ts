import set from "lodash/set"

import { TParsedProductImportRowData } from "./types"
import { csvRevertCellContentFormatter } from "../../../utils"

/**
 * Pick keys for a new object by regex.
 * @param data - Initial data object
 * @param regex - A regex used to pick which keys are going to be copied in the new object
 */
export function pickObjectPropsByRegex(
  data: TParsedProductImportRowData,
  regex: RegExp
): TParsedProductImportRowData {
  const variantKeyPredicate = (key: string): boolean => regex.test(key)
  const ret = {}

  for (const k in data) {
    if (variantKeyPredicate(k)) {
      ret[k] =
        typeof data[k] === "string"
          ? csvRevertCellContentFormatter(data[k] as string)
          : data[k]
    }
  }

  return ret
}

/**
 * Pick data from parsed CSV object relevant for product create/update and remove prefixes from keys.
 */
export function transformProductData(
  data: TParsedProductImportRowData
): TParsedProductImportRowData {
  const ret = {}
  const productData = pickObjectPropsByRegex(data, /product\./)

  Object.keys(productData).forEach((k) => {
    const key = k.split("product.")[1]
    set(ret, key, productData[k])
  })

  return ret
}

/**
 * Pick data from parsed CSV object relevant for variant create/update and remove prefixes from keys.
 */
export function transformVariantData(
  data: TParsedProductImportRowData
): TParsedProductImportRowData {
  const ret = {}
  const productData = pickObjectPropsByRegex(data, /variant\./)

  Object.keys(productData).forEach((k) => {
    const key = k.split("variant.")[1]
    set(ret, key, productData[k])
  })

  // include product handle to keep track of associated product
  ret["product.handle"] = data["product.handle"]
  set(ret, "product.options", data["product.options"])

  return ret
}
