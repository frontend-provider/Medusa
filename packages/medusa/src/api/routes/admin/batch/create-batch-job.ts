import { IsBoolean, IsObject, IsOptional, IsString } from "class-validator"

import { BatchJob } from "../../../../models"
import BatchJobService from "../../../../services/batch-job"
import { EntityManager } from "typeorm"
import { validator } from "../../../../utils/validator"

/**
 * @oas [post] /admin/batch-jobs
 * operationId: "PostBatchJobs"
 * summary: "Create a Batch Job"
 * description: "Creates a Batch Job."
 * x-authenticated: true
 * requestBody:
 *   content:
 *    application/json:
 *      schema:
 *        $ref: "#/components/schemas/AdminPostBatchesReq"
 * x-codegen:
 *   method: create
 * x-codeSamples:
 *   - lang: JavaScript
 *     label: JS Client
 *     source: |
 *       import Medusa from "@medusajs/medusa-js"
 *       const medusa = new Medusa({ baseUrl: MEDUSA_BACKEND_URL, maxRetries: 3 })
 *       // must be previously logged in or use api token
 *       medusa.admin.batchJobs.create({
 *         type: 'product-export',
 *         context: {},
 *         dry_run: false
 *       }).then((({ batch_job }) => {
 *         console.log(batch_job.id);
 *       });
 *   - lang: Shell
 *     label: cURL
 *     source: |
 *       curl --location --request POST 'https://medusa-url.com/admin/batch-jobs' \
 *       --header 'Content-Type: application/json' \
 *       --header 'Authorization: Bearer {api_token}' \
 *       --data-raw '{
 *           "type": "product-export",
 *           "context": { }
 *       }'
 * security:
 *   - api_token: []
 *   - cookie_auth: []
 * tags:
 *   - Batch Jobs
 * responses:
 *   201:
 *     description: OK
 *     content:
 *       application/json:
 *         schema:
 *           $ref: "#/components/schemas/AdminBatchJobRes"
 *   "400":
 *     $ref: "#/components/responses/400_error"
 *   "401":
 *     $ref: "#/components/responses/unauthorized"
 *   "404":
 *     $ref: "#/components/responses/not_found_error"
 *   "409":
 *     $ref: "#/components/responses/invalid_state_error"
 *   "422":
 *     $ref: "#/components/responses/invalid_request_error"
 *   "500":
 *     $ref: "#/components/responses/500_error"
 */
export default async (req, res) => {
  const validated = await validator(AdminPostBatchesReq, req.body)

  const batchJobService: BatchJobService = req.scope.resolve("batchJobService")

  const userId = req.user.id ?? req.user.userId

  const manager: EntityManager = req.scope.resolve("manager")
  const batch_job = await manager.transaction(async (transactionManager) => {
    const toCreate = await batchJobService
      .withTransaction(transactionManager)
      .prepareBatchJobForProcessing(validated, req)
    return await batchJobService.withTransaction(transactionManager).create({
      ...toCreate,
      created_by: userId,
    })
  })

  res.status(201).json({ batch_job })
}

/**
 * @schema AdminPostBatchesReq
 * type: object
 * required:
 *   - type
 *   - context
 * properties:
 *   type:
 *     type: string
 *     description: The type of batch job to start.
 *     example: product-export
 *   context:
 *     type: object
 *     description: Additional infomration regarding the batch to be used for processing.
 *     example:
 *       shape:
 *         prices:
 *           - region: null
 *             currency_code: "eur"
 *         dynamicImageColumnCount: 4
 *         dynamicOptionColumnCount: 2
 *       list_config:
 *         skip: 0
 *         take: 50
 *         order:
 *           created_at: "DESC"
 *         relations:
 *           - variants
 *           - variant.prices
 *           - images
 *   dry_run:
 *     type: boolean
 *     description: Set a batch job in dry_run mode to get some information on what will be done without applying any modifications.
 *     default: false
 */
export class AdminPostBatchesReq {
  @IsString()
  type: string

  @IsObject()
  context: BatchJob["context"]

  @IsBoolean()
  @IsOptional()
  dry_run = false
}
