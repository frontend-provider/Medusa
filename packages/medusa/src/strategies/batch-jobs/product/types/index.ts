import { BatchJob, Product, ProductVariant } from "../../../../models"
import { Selector } from "../../../../types/common"
import { CsvSchema, CsvSchemaColumn } from "../../../../interfaces/csv-parser"
import {
  BatchJobService,
  ProductCollectionService,
  ProductService,
  ProductVariantService,
  RegionService,
  SalesChannelService,
  ShippingProfileService,
} from "../../../../services"
import { FileService } from "medusa-interfaces"
import { FlagRouter } from "../../../../utils/flag-router"
import { EntityManager } from "typeorm"
import { IFileService } from "../../../../interfaces"

export type ProductExportInjectedDependencies = {
  manager: EntityManager
  batchJobService: BatchJobService
  productService: ProductService
  fileService: IFileService
  featureFlagRouter: FlagRouter
}

export type ProductExportBatchJobContext = {
  retry_count?: number
  max_retry?: number
  offset?: number
  limit?: number
  batch_size?: number
  order?: string
  fields?: string
  expand?: string
  shape: {
    prices: ProductExportPriceData[]
    dynamicOptionColumnCount: number
    dynamicImageColumnCount: number
    dynamicSalesChannelsColumnCount: number
  }
  list_config?: {
    select?: string[]
    relations?: string[]
    skip?: number
    take?: number
    order?: Record<string, "ASC" | "DESC">
  }
  filterable_fields?: Selector<unknown>
}

export type ProductExportBatchJob = BatchJob & {
  context: ProductExportBatchJobContext
}

export type ProductExportPriceData = {
  currency_code?: string
  region?: { name: string; currency_code: string; id: string }
}

export type ProductExportColumnSchemaEntity = "product" | "variant"

export type DynamicProductExportDescriptor = {
  isDynamic: true
  buildDynamicColumnName: (dataOrIndex: any) => string
}

export type ProductExportDescriptor =
  | {
      accessor: (product: Product) => string
      entityName: Extract<ProductExportColumnSchemaEntity, "product">
    }
  | {
      accessor: (variant: ProductVariant) => string
      entityName: Extract<ProductExportColumnSchemaEntity, "variant">
    }

export type ProductImportInjectedProps = {
  batchJobService: BatchJobService
  productService: ProductService
  productVariantService: ProductVariantService
  shippingProfileService: ShippingProfileService
  salesChannelService: SalesChannelService
  regionService: RegionService
  productCollectionService: ProductCollectionService
  fileService: typeof FileService

  featureFlagRouter: FlagRouter
  manager: EntityManager
}

/**
 * Import Batch job context column type.
 */
export type ProductImportJobContext = {
  total: number
  fileKey: string
}

export type ProductImportBatchJob = BatchJob & {
  result: Pick<BatchJob, "result"> & {
    operations: {
      [K in keyof typeof OperationType]: number
    }
  }
}

/**
 * Schema definition of for an import CSV file.
 */
export type ProductImportCsvSchema = CsvSchema<
  TParsedProductImportRowData,
  TBuiltProductImportLine
>

/**
 * Supported batch job import ops.
 */
export enum OperationType {
  ProductCreate = "PRODUCT_CREATE",
  ProductUpdate = "PRODUCT_UPDATE",
  VariantCreate = "VARIANT_CREATE",
  VariantUpdate = "VARIANT_UPDATE",
}

/**
 * Data shape returned by the CSVParser.
 */
export type TParsedProductImportRowData = Record<
  string,
  string | number | object | undefined | (string | number | object)[]
>

/**
 * CSV parser's row reducer result data shape.
 */
export type TBuiltProductImportLine = Record<string, any>

export type ProductImportDescriptor = CsvSchemaColumn<
  TParsedProductImportRowData,
  TBuiltProductImportLine,
  true
>

export type ProductColumnDefinition = {
  [key: string]: {
    name: string
    importDescriptor?: ProductImportDescriptor
    exportDescriptor?: ProductExportDescriptor | DynamicProductExportDescriptor
  }
}
