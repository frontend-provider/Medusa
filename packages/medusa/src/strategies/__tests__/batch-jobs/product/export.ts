import ProductExportStrategy from "../../../batch-jobs/product/export"
import { IdMap, MockManager } from "medusa-test-utils"
import { User } from "../../../../models"
import { BatchJobStatus } from "../../../../types/batch-job"
import { productsToExport } from "../../../__fixtures__/product-export-data"
import { AdminPostBatchesReq } from "../../../../api/routes/admin/batch/create-batch-job"
import { defaultAdminProductRelations } from "../../../../api/routes/admin/products"
import { ProductExportBatchJob } from "../../../batch-jobs/product"
import { Request } from "express"

const outputDataStorage: string[] = []

let fakeJob = {
  id: IdMap.getId("product-export-job"),
  type: 'product-export',
  created_by: IdMap.getId("product-export-job-creator"),
  created_by_user: {} as User,
  context: {},
  result: {},
  dry_run: false,
  status: BatchJobStatus.PROCESSING as BatchJobStatus
} as ProductExportBatchJob

const fileServiceMock = {
  delete: jest.fn(),
  getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
    return Promise.resolve({
      writeStream: {
        write: (data: string) => {
          outputDataStorage.push(data)
        },
        end: () => void 0
      },
      promise: Promise.resolve(),
      url: 'product-export.csv'
    })
  }),
  withTransaction: function () {
    return this
  }
}
const batchJobServiceMock = {
  withTransaction: function () {
    return this
  },
  update: jest.fn().mockImplementation((job, data) => {
    fakeJob = {
      ...fakeJob,
      ...data,
      context: { ...fakeJob?.context, ...data?.context },
      result: { ...fakeJob?.result, ...data?.result }
    }
    return Promise.resolve(fakeJob)
  }),
  updateStatus: jest.fn().mockImplementation((status) => {
    fakeJob.status = status
    return Promise.resolve(fakeJob)
  }),
  complete: jest.fn().mockImplementation(() => {
    fakeJob.status = BatchJobStatus.COMPLETED
    return Promise.resolve(fakeJob)
  }),
  retrieve: jest.fn().mockImplementation(() => {
    return Promise.resolve(fakeJob)
  }),
  setFailed: jest.fn().mockImplementation((...args) => {
    console.error(...args)
  })
}
const productServiceMock = {
  withTransaction: function () {
    return this
  },
  list: jest.fn().mockImplementation(() => Promise.resolve(productsToExport)),
  count: jest.fn().mockImplementation(() => Promise.resolve(productsToExport.length)),
  listAndCount: jest.fn().mockImplementation(() => {
    return Promise.resolve([productsToExport, productsToExport.length])
  }),
}
const managerMock = MockManager

describe("Product export strategy", () => {
  const productExportStrategy = new ProductExportStrategy({
    manager: managerMock,
    fileService: fileServiceMock as any,
    batchJobService: batchJobServiceMock as any,
    productService: productServiceMock as any,
  })

  it('should generate the appropriate template', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    const template = await productExportStrategy.buildHeader(fakeJob)
    expect(template).toMatch(/.*Product ID.*/)
    expect(template).toMatch(/.*Product Handle.*/)
    expect(template).toMatch(/.*Product Title.*/)
    expect(template).toMatch(/.*Product Subtitle.*/)
    expect(template).toMatch(/.*Product Description.*/)
    expect(template).toMatch(/.*Product Status.*/)
    expect(template).toMatch(/.*Product Thumbnail.*/)
    expect(template).toMatch(/.*Product Weight.*/)
    expect(template).toMatch(/.*Product Length.*/)
    expect(template).toMatch(/.*Product Width.*/)
    expect(template).toMatch(/.*Product Height.*/)
    expect(template).toMatch(/.*Product HS Code.*/)
    expect(template).toMatch(/.*Product Origin Country.*/)
    expect(template).toMatch(/.*Product MID Code.*/)
    expect(template).toMatch(/.*Product Material.*/)
    expect(template).toMatch(/.*Product Collection Title.*/)
    expect(template).toMatch(/.*Product Collection Handle.*/)
    expect(template).toMatch(/.*Product Type.*/)
    expect(template).toMatch(/.*Product Tags.*/)
    expect(template).toMatch(/.*Product Discountable.*/)
    expect(template).toMatch(/.*Product External ID.*/)
    expect(template).toMatch(/.*Product Profile Name.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)

    expect(template).toMatch(/.*Variant ID.*/)
    expect(template).toMatch(/.*Variant Title.*/)
    expect(template).toMatch(/.*Variant SKU.*/)
    expect(template).toMatch(/.*Variant Barcode.*/)
    expect(template).toMatch(/.*Variant Allow backorder.*/)
    expect(template).toMatch(/.*Variant Manage inventory.*/)
    expect(template).toMatch(/.*Variant Weight.*/)
    expect(template).toMatch(/.*Variant Length.*/)
    expect(template).toMatch(/.*Variant Width.*/)
    expect(template).toMatch(/.*Variant Height.*/)
    expect(template).toMatch(/.*Variant HS Code.*/)
    expect(template).toMatch(/.*Variant Origin Country.*/)
    expect(template).toMatch(/.*Variant MID Code.*/)
    expect(template).toMatch(/.*Variant Material.*/)

    expect(template).toMatch(/.*Option 1 Name.*/)
    expect(template).toMatch(/.*Option 1 Value.*/)
    expect(template).toMatch(/.*Option 2 Name.*/)
    expect(template).toMatch(/.*Option 2 Value.*/)

    expect(template).toMatch(/.*Price USD.*/)
    expect(template).toMatch(/.*Price france \[USD\].*/)
    expect(template).toMatch(/.*Price denmark \[DKK\].*/)
    expect(template).toMatch(/.*Price Denmark \[DKK\].*/)

    expect(template).toMatch(/.*Image 1 Url.*/)
  })

  it('should process the batch job and generate the appropriate output', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    await productExportStrategy.processJob(fakeJob.id)
    expect(outputDataStorage).toMatchSnapshot()
  })

  it('should prepare the job to be pre proccessed', async () => {
    const fakeJob1: AdminPostBatchesReq = {
      type: 'product-export',
      context: {
        limit: 10,
        offset: 10,
        expand: "variants",
        fields: "title",
        order: "-title",
        filterable_fields: { title: "test" }
      },
      dry_run: false
    }

    const output1 = await productExportStrategy.prepareBatchJobForProcessing(
      fakeJob1,
      {} as Express.Request
    )

    expect(output1.context).toEqual(expect.objectContaining({
      list_config: {
        select: ["title", "created_at", "id"],
        order: { title: "DESC" },
        relations: ["variants"],
        skip: 10,
        take: 10,
      },
      filterable_fields: { title: "test" }
    }))

    const fakeJob2: AdminPostBatchesReq = {
      type: 'product-export',
      context: {},
      dry_run: false
    }

    const output2 = await productExportStrategy.prepareBatchJobForProcessing(
      fakeJob2,
      {} as Express.Request
    )

    expect(output2.context).toEqual(expect.objectContaining({
      list_config: {
        select: undefined,
        order: { created_at: "DESC" },
        relations: [
          ...defaultAdminProductRelations,
          "variants.prices.region"
        ],
        skip: 0,
        take: 50,
      },
      filterable_fields: undefined
    }))
  })
})
