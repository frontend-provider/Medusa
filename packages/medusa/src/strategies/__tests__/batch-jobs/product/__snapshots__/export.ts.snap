// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product export strategy should process the batch job and generate the appropriate output 1`] = `
Array [
  "Product Id;Product Handle;Product Title;Product Subtitle;Product Description;Product Status;Product Thumbnail;Product Weight;Product Length;Product Width;Product Height;Product HS Code;Product Origin Country;Product MID Code;Product Material;Product Collection Title;Product Collection Handle;Product Type;Product Tags;Product Discountable;Product External Id;Product Profile Name;Product Profile Type;Variant Id;Variant Title;Variant SKU;Variant Barcode;Variant Inventory Quantity;Variant Allow Backorder;Variant Manage Inventory;Variant Weight;Variant Length;Variant Width;Variant Height;Variant HS Code;Variant Origin Country;Variant MID Code;Variant Material;Price france [USD];Price USD;Price denmark [DKK];Price Denmark [DKK];Option 1 Name;Option 1 Value;Option 2 Name;Option 2 Value;Image 1 Url
",
  "product-export-strategy-product-1;test-product-product-1;Test product;;\\"test-product-description-1
test-product-description-1 second line
test-product-description-1 third line
forth line\\";draft;;;;;;;;;;Test collection 1;test-collection1;test-type-1;123_1;true;;profile_1;profile_type_1;product-export-strategy-variant-1;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;1;1.1;1.3;;test-option-1;option 1 value 1;test-option-2;option 2 value 1;test-image.png
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;;;1.1;test-option;Option 1 value 1;;;test-image.png
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-3;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;1.2;;;test-option;Option 1 Value 1;;;test-image.png
",
]
`;

exports[`Product export strategy with sales Channels should process the batch job and generate the appropriate output 1`] = `
Array [
  "Product Id;Product Handle;Product Title;Product Subtitle;Product Description;Product Status;Product Thumbnail;Product Weight;Product Length;Product Width;Product Height;Product HS Code;Product Origin Country;Product MID Code;Product Material;Product Collection Title;Product Collection Handle;Product Type;Product Tags;Product Discountable;Product External Id;Product Profile Name;Product Profile Type;Variant Id;Variant Title;Variant SKU;Variant Barcode;Variant Inventory Quantity;Variant Allow Backorder;Variant Manage Inventory;Variant Weight;Variant Length;Variant Width;Variant Height;Variant HS Code;Variant Origin Country;Variant MID Code;Variant Material;Price france [USD];Price USD;Price denmark [DKK];Price Denmark [DKK];Option 1 Name;Option 1 Value;Option 2 Name;Option 2 Value;Image 1 Url;Sales Channel 1 Id;Sales Channel 1 Name;Sales Channel 1 Description;Sales Channel 2 Id;Sales Channel 2 Name;Sales Channel 2 Description
",
  "product-export-strategy-product-1;test-product-product-1;Test product;;\\"test-product-description-1
test-product-description-1 second line
test-product-description-1 third line
forth line\\";draft;;;;;;;;;;Test collection 1;test-collection1;test-type-1;123_1;true;;profile_1;profile_type_1;product-export-strategy-variant-1;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;1;1.1;1.3;;test-option-1;option 1 value 1;test-option-2;option 2 value 1;test-image.png;SC 1;SC 1;\\"SC 1
SC 1 second line
SC 1 third line
SC 1 forth line\\";;;
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;;;1.1;test-option;Option 1 value 1;;;test-image.png;SC 1;SC 1;SC 1;SC 2;SC 2;SC 2
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-3;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;1.2;;;test-option;Option 1 Value 1;;;test-image.png;SC 1;SC 1;SC 1;SC 2;SC 2;SC 2
",
]
`;
